brain_observatory_analysis/ophys/data_formatting.py

```diff
@@ -0,0 +1,533 @@
+    """Get a dataframe of traces from a session (all ophys trace, either dff or events) with no task
+    Assume 5 min gray screen before and after task
+    and 5 min fingerprint (movie-watching; 30 sec 10 iterations) imaging at the end
+    Some sessions don't have them.
+
+    Parameters
+    ----------
+    lamf_group : ExperimentGroup
+        ExperimentGroup object
+    session_name : str
+        session name
+    trace_type : str, optional
+        'dff' or 'events', by default 'dff'
+    
+    Returns
+    -------
+    trace_df_gray_pre_task : pd.DataFrame
+        dataframe of traces from gray screen before task
+    trace_df_gray_post_task : pd.DataFrame
+        dataframe of traces from gray screen after task
+    trace_df_fingerprint : pd.DataFrame
+        dataframe of traces from fingerprint (movie-watching; 30 sec 10 iterations) imaging at the end
+    """
+    gray_period = 5 * 60  # 5 minutes
+    oeids = np.sort(lamf_group.expt_table[lamf_group.expt_table.session_name==session_name].index.values)
+    stim_df = data_formatting.annotate_stimuli(lamf_group.experiments[oeids[0]])  # First experiment represents the session stimulus presentations
+    
+    # Match the # of indices from each experiment
+    start_inds = []
+    end_inds = []
+    post_gray_end_inds = []
+    for oeid in oeids:
+        timestamps = lamf_group.experiments[oeid].ophys_timestamps
+        
+        first_stim_start_time = stim_df.start_time.values[0]
+        if first_stim_start_time > gray_period - 30:  # 30 is an arbitrary buffer
+            run_pre_gray = True
+            first_stim_start_frame = np.where(timestamps > first_stim_start_time)[0][0]
+            start_inds.append(first_stim_start_frame)
+        else:
+            run_pre_gray = False
+
+        last_time = timestamps[-1]
+        last_stim_end_time = stim_df.stop_time.values[-1]
+        if last_time - last_stim_end_time > gray_period - 30:  # 30 is an arbitrary buffer
+            run_post_gray = True
+            last_stim_end_frame = np.where(timestamps > last_stim_end_time)[0][0]
+            end_inds.append(last_stim_end_frame)
+            post_gray_end_time = last_stim_end_time + gray_period
+
+            if last_time - post_gray_end_time > -60:  # -60 is an arbitrary buffer
+                run_fingerprint = True
+                post_gray_end_frame = np.where(timestamps > post_gray_end_time)[0][0]
+                post_gray_end_inds.append(post_gray_end_frame)
+            else:
+                run_fingerprint = False
+        else:
+            run_post_gray = False
+        
+    if run_pre_gray:
+        min_start_ind = np.min(start_inds)
+    if run_post_gray:
+        max_end_ind = np.max(end_inds)
+        min_post_gray_end_ind = np.min(post_gray_end_inds)
+
+    # Initialize
+    trace_df_gray_pre_task = pd.DataFrame(columns=['cell_specimen_id', 'trace', 'timepoints', 'oeid', 'target_region', 'depth_order', 'bisect_layer']).set_index('cell_specimen_id')
```

Comment:
looks like the array of columns is repeated throughout multiple functions, and the array is the same. One suggestion to make it either a function like "get_columns" to specify it just once. It's easier to make changes to the list if it's specified in one place. But it only makes sense if you think the list of columns is likely to be the same throughout all of the functions.

Reply:
Good point. Also, there will be cases where other columns are needed. I added functionality to give a list of column names as the input of the function.
I have thought about giving another function to set the columns, but not sure how general it should be. So I just left them as separate block of codes in different parts of the functions. How about getting back to it when there seems to be more general need for this?

Reply:
Actually, I added internal functions for default column names, default column base names, and setting column names.